### ISSUE TITLE
The inference of quantized models based on float16 weights fails on CPU 

### ISSUE THREAD

**Issue by dacorvo** [huggingface/quanto#39]
The inference of models with float16 weights in their Linear layers is not supported on CPU ('RuntimeError: "addmm_impl_cpu_" not implemented for 'Half').

Since addmm is not natively supported for integer either, we fallback to float16 for inference, which leads to a failure.

One possible workaround would be to fallback to float32, but it is a bit awkward (and it is not done for float models).

**Comment by shuokay** [comment]
Hi @dacorvo I suppose you are using torch as the inference framework according to the error message. I think it's might not a good idea to use torch directly.
1. torch is heavy to deploy
2. There are many other torch ops that didn't support int(int8, int16, int32 etc.) dtypes, it seems torch focus on training and float dtypes.
3. Torch lacks performance optimization for inference, especially quantized inference.

**Comment by dacorvo** [comment]
I don't understand your comment. quanto is a pytorch quantization toolkit.

**Comment by shuokay** [comment]
@dacorvo Sorry, I may have misunderstood the scope of Quanto. Previously, I thought Quanto was a fake quantization tool that exports fake-quantized or quantized models and then feeds them to inference frameworks like ONNXRuntime for inference. However, when I recalled that Quanto was implementing some inference operator kernels, I realized that Quanto itself also supports quantized inference.

One more question: What hardware will Quanto support for quantized inference in the short term? CUDA, x86, and mps?
